### PR TITLE
CPF-3261/SudoTicket

### PR DESCRIPTION
**Pull request by jlhughes**

### CODE REVIEW

File: src/foam/nanos/ruler/RuleEngine.java [hunk 1]
```diff
@@ -182,8 +182,8 @@ private void asyncApplyRules(List<Rule> rules, FObject obj, FObject oldObj) {
 
   private void retryAsyncApply(X x, Rule rule, FObject obj, FObject oldObj) {
     new RetryManager().submit(x, x1 -> {
-      rule.asyncApply(x, obj, oldObj, RuleEngine.this);
-      saveHistory(rule, obj);
+        rule.asyncApply(x, obj, oldObj, RuleEngine.this, rule);
```
Comment (TW80000) [inline src/foam/nanos/ruler/RuleEngine.java]: Indentation change wasn't necessary.

File: src/foam/nanos/ruler/RulerDAO.js [hunk 2]
```diff
@@ -164,7 +164,9 @@ return ret;`
       javaCode: `
 for ( Object key : sink.getGroupKeys() ) {
   RuleGroup rg = (RuleGroup) ((DAO) x.get("ruleGroupDAO")).find(key);
-  if( rg.f(x,obj,oldObj) ) {
+  if ( rg == null ) {
+   ((foam.nanos.logger.Logger) x.get("logger")).error("RuleGroup not found.", key);
```
Comment (TW80000) [inline src/foam/nanos/ruler/RulerDAO.js]:
```suggestion
    ((foam.nanos.logger.Logger) x.get("logger")).error("RuleGroup not found.", key);
```

File: src/foam/nanos/ruler/RulerDAO.js [hunk 3]
```diff
@@ -164,7 +164,9 @@ return ret;`
...
+  } else if( rg.f(x,obj,oldObj) ) {
```
Comment (TW80000) [inline src/foam/nanos/ruler/RulerDAO.js]:
```suggestion
  } else if ( rg.f(x, obj, oldObj) ) {
```

File: src/foam/nanos/ticket/Relationships.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  forwardName: 'tickets',
+  inverseName: 'owner',
+  cardinality: '1:*',
+  sourceDAOKey: 'userDAO',
+  unauthorizedSourceDAOKey: 'localUserDAO',
+  targetDAOKey: 'ticketDAO',
+  unauthorizedTargetDAOKey: 'localTicketDAO',
+  targetProperty: {
+    visibility: 'RO',
+//    section: '',
```
Comment (TW80000) [inline src/foam/nanos/ticket/Relationships.js]: There are a few bits of commented code in this file. They should be removed or uncommented.

File: src/foam/nanos/ticket/SudoTicket.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.ticket',
+  name: 'SudoTicket',
+  extends: 'foam.nanos.ticket.Ticket',
+
+  properties: [
+    {
+      name: 'user',
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicket.js]: How is `user` different from `owner`? I assume `owner` is the person who created the ticket and `user` is the user the owner wants to sudo as, but maybe that should be clarified by a doc comment on this property or by renaming `user` to something more specific. Maybe `targetUser` or `userToSudoAs`.

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  exends: 'foam.nanos.ruler.Rule',
+
+  implements: [
+    'foam.nanos.ruler.RuleAction'
+  ],
+
+  properties: [
+    {
+      name: 'description',
+      transient: true,
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: A transient description? Where does it get seen/used then?

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.ticket',
+  name: 'SudoTicketOpenRule',
+  exends: 'foam.nanos.ruler.Rule',
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: typo
```suggestion
  extends: 'foam.nanos.ruler.Rule',
```

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      transient: true,
+      visibility: 'RO',
+      value: ''
+    },
+    {
+      class: 'String',
+      name: 'ruleGroup',
+      value: 'sudo',
+      visibility: 'RO',
+      // readPermissionRequired: true,
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: Please resolve commented out code.

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    {
+      name: 'assignToGroup',
+      class: 'Reference',
+      of: 'foam.nanos.auth.Group',
+      value: 'admin',
+      view: function(_, x) {
+        return foam.u2.view.ChoiceView.create({
+          dao: x.groupDAO,
+          placeholder: 'Select... ',
+          // objToChoice: function(g) {
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: Ditto

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      nane: 'autoApprove',
+      value: false
+    },
+    {
+      name: 'assignToGroup',
+      class: 'Reference',
+      of: 'foam.nanos.auth.Group',
+      value: 'admin',
+      view: function(_, x) {
+        return foam.u2.view.ChoiceView.create({
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: I think the default view for reference properties is more appropriate. It should work out of the box if you set `targetDAOKey: 'groupDAO'` on this property.

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+          //   return [g.id, g.id];
+          // }
+        }, x);
+      },
+    },
+    {
+      name: 'action',
+      transient: true,
+      javaFactory: `
+    return this;
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]:
```suggestion
        return this;
```

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+      `
+    },
+    {
+      name: 'predicate',
+      javaFactory: `
+    return foam.nanos.ruler.predicate.IsInstancePredicate.Builder(x).setOf(foam.nanos.ticket.SudoTicket.getOwnClassInfo());
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]:
```suggestion
        return foam.nanos.ruler.predicate.IsInstancePredicate.Builder(x).setOf(foam.nanos.ticket.SudoTicket.getOwnClassInfo());
```

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+      `,
+      transient: true
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: Why is this transient?

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+          placeholder: 'Select... ',
+          // objToChoice: function(g) {
+          //   return [g.id, g.id];
+          // }
+        }, x);
+      },
+    },
+    {
+      name: 'action',
+      transient: true,
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: Why is this transient?

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+          public void execute(X x) {
+            Ticket ticket = (Ticket) obj.fclone();
+            DAO dao = ((DAO) x.get("approvalRequestDAO"))
+              .where(AND(
+                EQ(ApprovalRequest.DAO_KEY, "localTicketDAO"),
+                EQ(ApprovalRequest.OBJ_ID, ticket.getId())
+              ));
+
+            ApprovalStatus approval = ApprovalRequestUtil.getState(dao);
+            if ( approval != null && approval != ApprovalStatus.REQUESTED ) {
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: Shouldn't we have an `else` to handle when `approval` is null?

**Review by TW80000 — COMMENTED** [review]
Submitting what I have so far. Only got through 15/27 files so far. Will continue later.

File: src/foam/nanos/ticket/SudoTicketRule.js [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+              transaction.setStatus(
```
Comment (TW80000) [inline src/foam/nanos/ticket/SudoTicketRule.js]: Transaction? Where does this come from? Isn't this rule supposed to be related to sudo-ing and tickets?

Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 14 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 13 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 12 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 11 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 6 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 7 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 9 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 10 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicketRule.js (hunk 8 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicketRule.js]: Accidentally committed. Deleted.

File: src/foam/nanos/ticket/SudoTicket.js (hunk 5 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicket.js]: Strange, my copy is named 'sudoAsUser'.

File: src/foam/nanos/ruler/RuleEngine.java (hunk 1 above)
Comment (jlhughes) [reply src/foam/nanos/ruler/RuleEngine.java]: resolved.

File: src/foam/nanos/ticket/Relationships.js (hunk 4 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/Relationships.js]: updated.

File: src/foam/nanos/ticket/SudoTicket.js (hunk 5 above)
Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicket.js]: Also added doc.

Comment (jlhughes) [reply src/foam/nanos/ticket/SudoTicket.js]: This was old, moved to nanopay side.

**Pull request merged**